article: add part 3 of DReddit tutorial series (#138)
embark-framework/embark-site: source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md

That should do it! Let's spin up Embark, we should then see our component rendered on the screen:

<pre class="shell"><code>$ embark run</code></pre>

## Building a `CreatePost` component

Alright, enough warm up. Time to build components that are useful. We start off with building a component that lets users create posts through our application. Similar to `App`, we'll introduce a new component `createPost` that comes with a `render()` method to display a simple form for entering data. We'll also need to add event handlers to the form so that when a user submits the form, we can actually access the data and later on send it to our Smart Contract.

Creating a simple form is very straight forward:

<pre class="javascript"><code>import React, { Component } from 'react';

export class CreatePost extends Component {

  render() {
    return (
      &lt;form&gt;
        &lt;div&gt;
          &lt;label&gt;Topic&lt;/label&gt;
          &lt;input type="text" name="topic" /&gt;
        &lt;/div&gt;
        &lt;div&gt;
          &lt;textarea name="content"&gt;&lt;/textarea&gt;
        &lt;/div&gt;
        &lt;button&gt;Post&lt;/button&gt;
      &lt;/form&gt;
    )
  }
}
</code></pre>

> To actually render this component on screen, we need to make it part of our `App` component. Or, to be more specific, have the `App` component render our `CreatePost` component. For no we can simply extend `App`'s render function like this;

Author: 👍

Luckily, EmbarkJS comes with plenty convenient APIs to do exactly that! `EmbarkJS.Storage.saveText()` takes a string, uploads it to IPFS and returns its hash which can then be used to create a post using our Smart Contract. One thing to keep in mind is that those APIs are asynchronous. Similar to how we wrote tests in [part two](/news/2019/02/11/building-a-decentralized-reddit-with-embark-part-2/#Testing-createPost) of this tutorial, we'll use `async/await` to write asynchronous code in a synchronous fashion.

<pre class="javascript"><code>async createPost(event) {
  event.preventDefault();
  
  this.setState({
    loading: true
  });

  const ipfsHash = await EmbarkJS.Storage.saveText(JSON.stringify({
    topic: this.state.topic,
    content: this.state.content
  }));

  this.setState({
    topic: '',
    content: '',
    loading: false
  });
}</code></pre>

> We use `JSON.stringify()` on an object that holds the `topic` and `content` of the post to be created. This is also the first time we put `loading` into action. Setting it to `true` before, and `false` after we've performed our operations lets us render a useful message to the user as she's waiting for updates.

Author: Ha, and here I thought I'd do something good for being inclusive :D

> We use `JSON.stringify()` on an object that holds the `topic` and `content` of the post to be created. This is also the first time we put `loading` into action. Setting it to `true` before, and `false` after we've performed our operations lets us render a useful message to the user as she's waiting for updates.

Reviewer: Haha, yeah, I didn't want to sound like a cis-male. Just that it's the only place we use `she` so it felt random. I change dthe sentence so use neither so we don't frustrate anyone haha

### Sending transactions to create posts

To send a transaction to our Smart Contract, we can again take advantage of EmbarkJS' APIs, similar to how we did it in the [second part](/news/2019/02/11/building-a-decentralized-reddit-with-embark-part-2). We also need to get hold of an Ethereum account to send the transaction from. This will be very straight forward as we'll be just relying on the accounts that are generated by the Ethereum node that Embark spins up for us.

Once we have those things in place we can get a gas estimation for our transaction and send the data over. Here's how we retrieve our accounts, notice that `async/await` can be used here as well:

<pre class="javascript"><code>async createPost(event) {
  ...
  const accounts = await web3.eth.getAccounts();
  ...
}</code></pre>

Next up we'll import a `DReddit` Smart Contract instance from EmbarkJS and use it to get a gas estimation from `web3`. We can then use the estimation and one of our accounts to actually send the transaction:

<pre class="javascript"><code>import DReddit from 'Embark/contracts/DReddit';
...

async createPost(event) {
  ...
  const accounts = await web3.eth.getAccounts();

>   const createPost = await DReddit.methods.createPost(web3.utils.toHex(ipfsHash));

Author: I think I've addressed this in the paragraph before that code snippet

Let's stay consistent with our naming and say the property we're expecting to be filled with data is called `description`, just like the one used inside the Smart Contract. We can then use `EmbarkJS.Storage.get()` with the IPFS hash to fetch the data that represents the actual post. In order to render the data inside `Post`'s view, we'll parse it and use `setState()` accordingly.

To make sure all of that happens once the component is ready doing its work, we'll do all of that inside its `componentDidMount()` life cycle hook:

<pre class="javascript"><code>import React, { Component } from 'react';

export class Post extends Component {

  constructor(props) {
    super(props);
  }

  async componentDidMount() {
    const ipfsHash = web3.utils.toAscii(this.props.description);

>     const data = await EmbarkJS.Storage.get(ipfsHash);

Author: 👍 good catcH!

We can already put that in our `App` component. It won't render anything as we haven't fetched any posts yet, but that we'll do next.

<pre class="javascript"><code>import { List } from './List';

export class App extends Component {

  render() {
    return (
      &lt;React.Fragment&gt;
        &lt;h1&gt;DReddit&lt;/h1&gt;
        &lt;CreatePost /&gt;
        &lt;List /&gt;
      &lt;/React.Fragment&gt;
    )
  }
}
</code></pre>

### Fetching posts data

Let's fill our new `List` component with life! As mentioned earlier, we'll use our Smart Contract's `numPosts()` method to get hold of the total number of posts available. We then use that number to iterate over all indices and request every post individually. Since this is logic we want to execute once `List` component is ready, we'll use its `componentDidMount()` method for that:

<pre class="javascript"><code>export class List extends Component {
  ...
  async componentDidMount() {
    const totalPosts = await DReddit.methods.numPosts().call();

    let list = [];

>     if (totalPosts > 0) {

Author: Yes!

        const post = DReddit.methods.posts(i).call();
        list.push(post);
      }
    }

    list = await Promise.all(list);
  }
  ...
}</code></pre>

Notice that in the above code we don't `await` the calls to every individual post. This is on purpose as we don't want to wait on each and every promise to resolve, but first collect all of the promises we need and then resolve them all in one go using `Promise.all().`

Last but not least, we need to add an `id` property to every post as mentioned earlier. This is easily done by simply iterating over all posts and assigning the post's index as `id`. Once that is done, we can use `setState()` to update our component's state and render the list:

<pre class="javascript"><code>async componentDidMount() {
  ...
  list = list.map((post, index) => {
    post.id = index;
    return post;
  });

  this.setState({ posts: list });
}</code></pre>

That's it! Our application now renders a list of all created posts. Unfortunately, posts are not being re-fetched  automatically when adding new posts. For the time being, we'll have to reload the browser every time after adding a post. However, this we'll address now.

### Reloading posts

> There is certainly different ways to make the list of posts update automatically, so take the following approach with a grain of salt. What we need is a way to have the `createPost` component tell the `List` component to reload its posts. However, there's not communication layer in place when building a simple React app like this, so the most straight forward way to make this possible, is to move the logic of loading the posts in the parent component of `CreatePost` and `List` (in our case `App`), and have it pass that logic down to places where its needed. This also means we'll be fetching the list inside `App` and pass down the pure data to `List`.

Reviewer: ```suggestion
There is certainly different ways to make the list of posts update automatically, so take the following approach with a grain of salt. What we need is a way to have the `createPost` component tell the `List` component to reload its posts. However, there's no communication layer in place when building a simple React app like this, so the most straight forward way to make this possible, is to move the logic of loading the posts in the parent component of `CreatePost` and `List` (in our case `App`), and have it pass that logic down to places where its needed. This also means we'll be fetching the list inside `App` and pass down the pure data to `List`.
```

Reviewer: Also, I just thought, there is another way to do it. You can always use contract `events` in the List and listen to new posts.

Author: Yea good point!

If this sounds overwhelming, no worries, it's more trivial than that! Let's start by introducing a `loadPosts()` function in our `App` component. Essentially we're moving everything from `List`'s `componentDidMount()` function into `App`:

<pre class="javascript"><code>export class App extends Component {
  ...
  async loadPosts() {
    const totalPosts = await DReddit.methods.numPosts().call();

    let list = [];

    if (totalPosts > 0) {
      for (let i = 0; i < totalPosts; i++) {
        const post = DReddit.methods.posts(i).call();
        list.push(post);
      }
    }

    list = await Promise.all(list);
    list = list.map((post, index) => {
      post.id = index;
      return post;
    });

    list;

    this.setState({ posts: list });
  }
}</code></pre>

To make this work we also need to introduce a `state` with the dedicated `posts`. After that, we make sure `loadPosts()` is called when `App` is mounted:

<pre class="javascript"><code>export class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      posts: []
    };
  }

  async componentDidMount() {
    await this.loadPosts();
  }
  ...
}</code></pre>

> Last but not least, all we have to do is passing the `posts` down to `List` and `loadPosts()` to `CreatePost` as a callback handler if you will:

Reviewer: You use "Last but not least" a lot. Maybe changing it from time to time.

Reviewer: ```suggestion
Last but not least, all we have to do is pass the `posts` down to `List` and `loadPosts()` to `CreatePost` as a callback handler if you will:
```

<pre class="javascript"><code>render() {
  return (
    &lt;React.Fragment&gt;
      &lt;h1&gt;DReddit&lt;/h1&gt;
      &lt;CreatePost afterPostHandler={this.loadPosts.bind(this)}/&gt;
      &lt;List posts={this.state.posts}/&gt;
    &lt;/React.Fragment&gt;
  )
}
</code></pre>

> Once that is done, we can consume `posts` and `afterPostHandler()` from `this.props` respectively. In `List` component's `render()` function we'll do (notice we don't rely on `this.state` anymore):

Reviewer: ```suggestion
Once that is done, we can consume `posts` and `afterPostHandler()` from `this.props` respectively. In `List`'s `render()` function we'll do (notice we don't rely on `this.state` anymore):
```

<pre class="javascript"><code>async loadPosts() {
  ...
  list = list.map((post, index) => {
    post.id = index;
    post.upvotes = parseInt(post.upvotes, 10);
    post.downvotes = parseInt(post.downvotes, 10);
    return post;
  });
  ...
}</code></pre>

Once that is done we can pass each post's `upvotes` and `downvotes` to every `Post` component via its `props` inside our `List` component:

<pre class="javascript"><code>export class List extends Component {
  ...
  render() {
    return (&lt;React.Fragment&gt;
      {this.props.posts.map(post => {
        return (&lt;Post 
          key={post.id}
          description={post.description}
          creationDate={post.creationDate}
          upvotes={post.upvotes}
          downvotes={post.downvotes}
          owner={post.owner}
          /&gt;)
      })}
      &lt;/React.Fragment&gt;
    )
  }
}</code></pre>

> Rendering the number of `upvotes` and `downvotes` is then really just a matter of interpolating them in `Post` component's `render()` function. We'll just going to add them next to the buttons, but feel free to put them somewhere else:

Reviewer: ```suggestion
Rendering the number of `upvotes` and `downvotes` is then really just a matter of interpolating them in `Post`'s `render()` function. We're just going to add them next to the buttons, but feel free to put them somewhere else:
```
